server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1416,29 +1416,29 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
 
         if (source.slice() != null) {
             if (source.pointInTimeBuilder() == null && context.scrollContext() == null) {
-                throw new SearchException(shardTarget, "[slice] can only be used with [scroll] or [point-in-time] requests");
+                throw new SearchUsageException(shardTarget, "[slice] can only be used with [scroll] or [point-in-time] requests");
             }
             context.sliceBuilder(source.slice());
         }
 
         if (source.storedFields() != null) {
             if (source.storedFields().fetchFields() == false) {
                 if (context.sourceRequested()) {
-                    throw new SearchException(shardTarget, "[stored_fields] cannot be disabled if [_source] is requested");
+                    throw new SearchUsageException(shardTarget, "[stored_fields] cannot be disabled if [_source] is requested");
                 }
                 if (context.fetchFieldsContext() != null) {
-                    throw new SearchException(shardTarget, "[stored_fields] cannot be disabled when using the [fields] option");
+                    throw new SearchUsageException(shardTarget, "[stored_fields] cannot be disabled when using the [fields] option");
                 }
             }
             context.storedFieldsContext(source.storedFields());
         }
 
         if (source.collapse() != null) {
             if (context.scrollContext() != null) {
-                throw new SearchException(shardTarget, "cannot use `collapse` in a scroll context");
+                throw new SearchUsageException(shardTarget, "cannot use `collapse` in a scroll context");
             }
             if (context.rescore() != null && context.rescore().isEmpty() == false) {
-                throw new SearchException(shardTarget, "cannot use `collapse` in conjunction with `rescore`");
+                throw new SearchUsageException(shardTarget, "cannot use `collapse` in conjunction with `rescore`");
```

Comment:
As I was reviewing this one more time, I have been wondering: do these exceptions need to be thrown in the shards? Aren't these situations that should / could be detected earlier on in the coordinating node? That would remove the need to associate the exceptions with a specific shard target, and may allow us to just throw an IllegalArgumentException in the coordinator? Am I missing something? Do we ever test these error messages?

Reply:
These are all occurring in `SearchService.parseSource` method. The SearchServiceTests method calls that method (via createContext) in many of its tests. I put a breakpoint at the spots where SearchUsageException is created and ran the tests - no breakpoints were hit when running in the debugger. So no, these don't have any test coverage.

> Aren't these situations that should / could be detected earlier on in the coordinating node? 

It needs a SearchContext to determine some of these usage errors, so if that context can be created on the coordinator before it starts the search phases, then yes.

But to create a context you need:

```java
    DefaultSearchContext(
        ReaderContext readerContext,
        ShardSearchRequest request,
        SearchShardTarget shardTarget,
        LongSupplier relativeTimeSupplier,
        TimeValue timeout,
        FetchPhase fetchPhase,
        boolean lowLevelCancellation,
        Executor executor,
        SearchService.ResultsType resultsType,
        boolean enableQueryPhaseParallelCollection,
        int minimumDocsPerSlice
```

So it is designed to happen at the shard level.

Reply:
I am thinking that we should not necessarily rely on the search context for this validation. The info that we consult are all taken from the incoming request, I believe. But this would require a bit of rework, there may be some massaging that we do on the incoming request parameters and what we get on the shards is not exactly always what the user provided. That does not feel like a good reason to perform validation on the shards that we could perform on the coordinating node. If you don't get to this, I am happy to take this over next and see how far I get with it. In the worst case, we go back to your suggested approach. Thanks for the work on this.

Reply:
OK, since you are the next to take on the "operational excellence" rotation, you could take over this work and try to move the validation to the coordinator. I'll be happy to review, so I can learn from it.